Added support for overriding the EKS Endpoint

treacher opened the pull request:
### Description

Added support for overriding the EKS Endpoint for the following commands:

- `eksctl get iamidentitymapping`
- `eksctl create iamidentitymapping`
- `eksctl delete iamidentitymapping`
- `eksctl create iamserviceaccount`
- `eksctl delete iamserviceaccount`
- `eksctl create nodegroup`
- `eksctl delete nodegroup`
- `eksctl drain nodegroup`
- `eksctl delete cluster`

Updated the Kubeconfig class so that it accepts an override for the EKS Endpoint.

Refactored explicit arguments into a params object. I followed the pattern used in pkg/ctl/create/cluster.go.

Added an explicit `--cluster-endpoint-override` flag for defining the EKS Endpoint you'd like to use for the eksctl commands.

This change allows users with a private cluster to be able to use SSH tunnelling to access the cluster from their local machine via a specified port.

Issue #1552 

During the making of this change, I had considered that perhaps it makes more sense to allow a port override instead of the entire cluster URL. That being said changing the full URL gives you a bit more power if for example you decide to put your cluster endpoint behind a DNS record.

### Manual testing performed:
Ran the following commands with and without passing in the `--cluster-endpoint-override` flag:
- `eksctl get iamidentitymapping`
- `eksctl create iamidentitymapping`
- `eksctl delete iamidentitymapping`
- `eksctl create iamserviceaccount`
- `eksctl delete iamserviceaccount`
- `eksctl create nodegroup`
- `eksctl delete nodegroup`
- `eksctl drain nodegroup`
- `eksctl delete cluster`

1. I observed that the above commands were still using the EKSEndpoint output from the EKS Cluster stack when not passing in the `--cluster-endpoint-override` flag and that the command succeeded.
2. I observed that the above commands used my specified EKS Endpoint when passing the `--cluster-endpoint-override` flag and that the command succeeded.

### Checklist
- [x] Added tests that cover your change (if possible)
- [x] Added/modified documentation as required (such as the `README.md`, and `examples` directory)
- [x] Manually tested

rndstr commented on pkg/ctl/cmdutils/cmdutils.go:
```diff
@@ -173,6 +173,11 @@ func AddUpdateAuthConfigMap(fs *pflag.FlagSet, updateAuthConfigMap *bool, descri
 
+// AddClusterEndpointOverrideFlag adds common cluster-endpoint-override flag
+func AddClusterEndpointOverrideFlag(fs *pflag.FlagSet, clusterEndpoint *string) {
+	fs.StringVar(clusterEndpoint, "cluster-endpoint-override", "", "cluster endpoint to use instead of the eks cluster endpoint provided by Cloudformation")
```
it feels like we wouldn't need the `-override` suffix since cli flags generally take precedence anyway. i would just use `--cluster-endpoint` here.

treacher replied:
Yeah, very good point.

jitran commented:
This change looks sound to me. Nice use of structs to improve the readability :)

Should we raise a new PR to deprecate the AWS_EKS_ENDPOINT environment variable?

treacher commented:
Thanks for your thorough review @rndstr, I've done the following:
- Fixed the grammar errors
- Rebased against master (this required one change as a NewClient call was made with the old interface [here](https://github.com/weaveworks/eksctl/blob/f483e1bace0e27220f42d9f5448939272728237f/pkg/ctl/create/fargate.go)
- Updated the CLI flag to `--cluster-endpoint`

I'll await a decision in regards to adding support in the ClusterConfig. :)

martina-if commented:
Hi @treacher , thank you for the PR! I am curious as I haven't used it myself yet: does setting the environment variable `AWS_EKS_ENDPOINT` to the specified endpoint url work for these cases? Was it only working for specific cases and not others?

jitran commented:
Hi @martina-if 

Here's what I tried:

Add the cluster's ca chain to the mac key chain so that the cli tools can trust the cluster endpoint
```
aws eks describe-cluster --name <cluster-name> --query cluster.certificateAuthority.data | xargs | base64 -D > ca.pem
sudo security add-trusted-cert -d -r trustRoot -k /Library/Keychains/System.keychain ca.pem

# create the ssh tunnel and test the endpoint
ssh -L 8001:<cluster-endpoint>:443 <cluster-name-ssh-config>
curl -v https://<cluster-endpoint>:8001/clusters/<cluster-name>
```

Now attempting to use the environment export:
```
export AWS_EKS_ENDPOINT=https://<cluster-endpoint>:8001
eksctl delete nodegroup --name <node-group-name> --cluster <cluster-name>
```

It gets pass the untrusted ssl certificate trust issue, but fails with:
```
[✖]  fetching cluster status to determine operability: unable to describe cluster control plane: SerializationError: failed to unmarshal response error
    status code: 403, request id:
caused by: UnmarshalError: failed decoding error message
HEX DUMP (REDACTED) |{"kind":"Status"|
HEX DUMP (REDACTED) |,"apiVersion":"v|
HEX DUMP (REDACTED) |1","metadata":{}|
HEX DUMP (REDACTED) |,"status":"Failu|
HEX DUMP (REDACTED) |re","message":"f|
HEX DUMP (REDACTED) |orbidden: User \|
HEX DUMP (REDACTED) |"system:anonymou|
HEX DUMP (REDACTED) |s\" cannot get p|
HEX DUMP (REDACTED) |ath \"/clusters/|
HEX DUMP (REDACTED) |<cluster-name>\"|
HEX DUMP (REDACTED) |","reason":"Forb|
HEX DUMP (REDACTED) |idden","details"|
HEX DUMP (REDACTED) |:{},"code":403}.|

caused by: json: cannot unmarshal number into Go struct field jsonErrorResponse.code of type string
```

It seems as though the aws-iam-authenticator token is not passed correctly to the endpoint.

martina-if commented:
I guess that value of the variable is not passed to the kubeconfig properly (like you correctly do here for the flag https://github.com/weaveworks/eksctl/pull/1618/files#diff-811c261701d0690beed656bfd959b131R72). Do you think you can modify this PR to pass the endpoint all the way from the env variable instead of the flag?

jitran commented:
> I guess that value of the variable is not passed to the kubeconfig properly (like you correctly do here for the flag https://github.com/weaveworks/eksctl/pull/1618/files#diff-811c261701d0690beed656bfd959b131R72). Do you think you can modify this PR to pass the endpoint all the way from the env variable instead of the flag?

Hi @martina-if, I couldn't find any documentation for the AWS_EKS_ENDPOINT environment variable? Is that a widely used option?

I personally would prefer to explicitly define the cluster api via an eksctl argument or inside the config file. This follows the current pattern for the existing eksctl arguments.

treacher commented:
> I guess that value of the variable is not passed to the kubeconfig properly (like you correctly do here for the flag https://github.com/weaveworks/eksctl/pull/1618/files#diff-811c261701d0690beed656bfd959b131R72). Do you think you can modify this PR to pass the endpoint all the way from the env variable instead of the flag?

I believe the flag `--cluster-endpoint` and the environment variable `AWS_EKS_ENDPOINT` have very different use cases.

`AWS_EKS_ENDPOINT` seems to be used for overwriting the endpoint that's used for the EKS API calls, see: https://github.com/aws/aws-sdk-go/blob/master/service/eks/service.go#L59 that's why @jitran will be seeing errors.

`--cluster-endpoint` is used for overwriting the endpoint that's used in the generated kube config file.

If we pass `foobar.amazon.com:8001` as the AWS_EKS_ENDPOINT then the AWS EKS API calls will fail as the actual cluster url is `foobar.amazon.com` not `foobar.amazon.com:8001`.

We have a couple of options here:
1. We use the changes defined here.
2. We create a new environment variable that can be set like `KUBECONFIG_CLUSTER_ENDPOINT`

Option 1 will be a bit more complex as I'll need to rebase etc and there are a lot of code changes that'll be merged. It's also a bit annoying to have to explicitly define the cluster endpoint on each CLI call.

Option 2 is probably going to be the easiest from an implementation complexity point of view but there then needs to be some documentation added around this variable.

@martina-if let me know what you think :)

martina-if commented:
Hi @treacher , you are right, I confused both endpoints. 

What you say makes sense to me: having an environment variable means the user doesn't need to type it every time. I guess this value could be written into the kubeconfig and then picked up by the kube client for subsequent calls after the cluster has been created. I would say go for it!

treacher commented:
I'm going to create a new PR and link it to this PR just to reduce confusion as the suggested change is quite different from the original.
